### PR TITLE
KAFKA-5277: Sticky Assignor should not cache previous assignment (KIP-54 follow-up)

### PR DESCRIPTION
**Pull request by vahidhashemian**
... plus some minor cleanup

### CODE REVIEW

**Comment by vahidhashemian** [comment]
@hachikuji For your review. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java [hunk 1]
```diff
@@ -186,15 +186,16 @@
     private static final Schema STICKY_ASSIGNOR_USER_DATA = new Schema(
             new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT)));
 
-    Map<String, List<TopicPartition>> currentAssignment = new HashMap<>();
+    Map<String, List<TopicPartition>> currentAssignment;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java]: Hmm.. if we're not caching the previous assignment, why do we still need this field?

Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java]: I kept it as it's being used in almost all the methods. But I guess I could define it locally in `assign(...)` and then pass it when necessary. Would that make sense?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java]: Yes, that would make sense to me.

File: clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java [hunk 2]
```diff
@@ -519,9 +524,14 @@ private boolean canParticipateInReassignment(String consumer,
...
-    private void balance(List<TopicPartition> sortedPartitions, List<TopicPartition> unassignedPartitions, TreeSet<String> sortedCurrentSubscriptions,
-                         HashMap<String, List<TopicPartition>> consumer2AllPotentialPartitions, HashMap<TopicPartition, List<String>> partition2AllPotentialConsumers,
-                         Map<String, List<TopicPartition>> oldAssignment, HashMap<TopicPartition, String> currentPartitionConsumer) {
+    private void balance(Map<String, List<TopicPartition>> currentAssignment,
+                         List<TopicPartition> sortedPartitions,
+                         List<TopicPartition> unassignedPartitions,
+                         TreeSet<String> sortedCurrentSubscriptions,
+                         Map<String, List<TopicPartition>> consumer2AllPotentialPartitions,
+                         Map<TopicPartition, List<String>> partition2AllPotentialConsumers,
+                         Map<String, List<TopicPartition>> oldAssignment,
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java]: nit: this is unused

Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java]: Thanks. You're right. Will remove it.

**Comment by hachikuji** [comment]
@vahidhashemian Thanks for the updates. LGTM

**Comment by vahidhashemian** [comment]
@hachikuji, could you please take a look at [KAFKA-7026](https://issues.apache.org/jira/browse/KAFKA-7026)? It seems to me that having the leader keep the latest assignment might be useful after all (as a potential and simple fix for the reported bug). Will be happy to explain further if needed. Thanks!

**Pull request closed without merging**